app/javascript/packages/config/get-config-value.ts

```diff
@@ -0,0 +1,36 @@
+/**
+ * Supported page configuration values.
+ */
+export interface Config {
+  /**
+   * Application name.
+   */
+  appName: string;
+
+  /**
+   * URL for analytics logging endpoint.
+   */
+  analyticsEndpoint: string;
+}
+
+/**
+ * Cached configuration.
+ */
+let config: Config;
+
+/**
+ * Returns the value associated as initialized through page configuration, if available.
+ *
+ * @param key Key for which to return value.
+ *
+ * @return Value, if exists.
+ */
+function getConfigValue<K extends keyof Config>(key: K): Config[K] {
+  if (config === undefined) {
+    config = JSON.parse(document.querySelector('[data-config]')?.textContent || '');
+  }
```

Comment:
yay big fan of caching it like this -- would it also make sense to add a method to clear/reset the config value for tests and such?

Reply:
> yay big fan of caching it like this -- would it also make sense to add a method to clear/reset the config value for tests and such?

Yeah, turns out we'll need some way to reset it in tests. I went through a few different ideas for how to implement this, and landed at the changes in 7378e6a to exempt caching in the test environment. I had dabbled with a few ideas including resetting via method or property, but the downside is that either each spec file which uses a config value would need to be aware to do the reset, or apply it globally in test setup, neither of which were very appealing to me.

The nice thing about `NODE_ENV` conditions is that through Webpack source replacement + Terser dead code elimination, the logic gets excluded from production build.

- https://webpack.js.org/plugins/environment-plugin/
- https://try.terser.org/

```js
let config;
const isCacheEnvironment = 'production' !== 'test';
function getValue(key) {
  if (config === undefined || !isCacheEnvironment) {
    config = {};
  }
  return config[key];
}

console.log(getValue('example'));
```

```js
let e;console.log((void 0===e&&(e={}),e["example"]));
```

Reply:
That seems like a good compromise! 👍